internal/retry/sizer.go

```diff
@@ -0,0 +1,60 @@
+package retry
+
+// Hint is a type for retry hints
+type Hint interface {
+	size() int
+}
+
+type hintInt int
+
+func (h hintInt) size() int {
+	return int(h)
+}
+
+// DoubleSize is a hint to retry with double the size
+const DoubleSize = hintInt(0)
+
+// Size returns a hint for a specific size
+func Size(s int) Hint {
+	return hintInt(s)
+}
+
+// SizeFunc is used to implement 'resize loops' that hides the complexity of the
+// sizing away from most of the application. It's a function that takes a size
+// argument and returns nil, if no retry is necessary, or a hint indicating the
+// size for the next retry. If errors or other results are required from the
+// function, the function can write them to function closures of the surrounding
+// scope. See tests for examples.
+type SizeFunc func(size int) (hint Hint)
+
+// WithSizes repeatingly calls a SizeFunc with increasing sizes until either it
+// returns nil, or the max size has been reached. If the returned hint is
+// DoubleSize or indicating a size not greater than the current size, the size
+// is doubled. If the hint or next size is greater than the max size, the max
+// size is used for a last retry.
+func WithSizes(size int, max int, f SizeFunc) {
+	for {
+		hint := f(size)
+		if hint == nil || size == max {
```

Comment:
Just to be a tad bit defensive against bugs outside of our control: can we make this a `size >= max`?

Reply:
Oh, actually you pointed me to something that I consider a bug in this code. The last line of the for loop is setting it to max in case it's larger, but I don't check that for the first loop.